### DISCUSSION TITLE
Migrating from pybricks v2.0 to version 3.3.x on EV3

### DISCUSSION THREAD

**Discussion by AAATom** [pybricks/support#1079] [General]
Intro: I'm co-coaching my son's school team which is preparing for national WRO competition. The v2.0 works fine up to its limits, one of largest being that `DriveBase` has `Stop.HOLD` as default and no way to change it (like using `then` parameter in v3.x), what makes movements jerky. I assume this could be worked around with some clever code but, hey, they are just kids starting with Python and "robotics" and would be happy to see such a thing working out-of -the-box (like they found it in the docs, alas, for v3.2).

Last week I have finally found the eluding EV3 "pybricks-micropython" executable (thanks  David!) and have successfully installed it (cookbook [here](https://github.com/orgs/pybricks/discussions/1075#discussioncomment-5976893).

It works well so far apart from some issues, which I intend to post here in this thread and, hopefully, get some constructive responses from the community :slightly_smiling_face:

**Comment by AAATom** [comment]
### Proportional line follower gone bad
The first issue and currently the deal breaking one for the team (in using v3.3.) is that line follower code that works fine in v2.0 works badly in v3.3 (build 2828).
When starting away from the line (over white) it cannot detect the line (or rather, the line edge) and wanders off to the other side of the line and on.

With some debug prints we discovered that the loop runs some 3 times faster on v3.3 (makes 3 times more passes) for the same distance the `DriveBase` covers.
The values read by the `ColorSensor` are practically the same - except that at the point where it gets over the line edge it does not result in the same behavior.

The line follower code is this:
```python
def TrackLineDistance(robot:DriveBase, distance:int, speed:int,
                      sensor:ColorSensor, isLeftSide:bool,
                      sensorThreshold:int, proportionalGain:float):
    
    sign = -1
    if isLeftSide:
        sign = 1
    
    robot.reset
    
    while robot.distance() < distance:
        deviation = sensor.reflection() - sensorThreshold
        turn_rate = proportionalGain * deviation * sign
        robot.drive(speed, turn_rate)
```

The parameters pass to `TrackLineDistance()` are
   speed = 40
   sensorThreshold = 20
   proportionalGain = 2

The team tried changing the speed and gain but it did not change much.
At the moment they are back to v2.0 as they don't have much time to debug this.

Here is a graph of the v2.0 data:
![Proportional line follower  on pybricks v2.0](https://i.postimg.cc/dtn2pGmG/image.png)

And graph of v3.3 data:
![Proportional line follower  on pybricks v3.3](https://i.postimg.cc/9fvTjpnN/image.png)

Notice the "hump" in v3.3 that is not there in v2.0 - it is the robot passing over the other edge of the line into white area and then into blue area where it wanders off mostly straight.

**What could be the reason v3.3. is behaving so differently?**

**Reply by laurensvalk** [reply]
I think this is the same as https://github.com/pybricks/support/issues/1035.

This is why we currently don't release EV3 versions and recommend that people use 2.0 for the time being :)

But we do get quite a lot of questions from WRO teams about this, so hopefully we can pick this up at some point.

**Reply by AAATom** [reply]
Thanks! I am aware about potential problems with v3.x on EV3, of course. 

Reading posts in #1035, looks like these things were appearing on and off after v3.2 - **can I get that version please?**

Interesting is that other driving code worked just fine. Is this something about doing `robot.drive()` in a loop?
Should I try adding some delays with `wait()` as mentioned in couple of places?

**Reply by laurensvalk** [reply]
Yes, that’s always good.
